File: 5Days-SoftwareArchitecture/Java/Day-1 CC Metrics/orderCboDemo.ts
class Logger {
    log(msg: string) { console.log("[Logger] " + msg); }
}
class SqlDatabase {
    save(payload: string) { console.log("[SqlDatabase] saved: " + payload); }
}
class LegacyDb {
    persist(payload: string) { console.log("[LegacyDb] persisted: " + payload); }
}
class LegacyDbAdapter extends SqlDatabase {
    private legacy = new LegacyDb();
    save(payload: string) { this.legacy.persist(payload); }
}
type Order = { id: string, amount: number };

class OrderService {
    checkout(o: Order) {
        const logger = new Logger();
        const db = new SqlDatabase();
        logger.log("Processing " + JSON.stringify(o));
        db.save(JSON.stringify(o));
        logger.log("Completed " + JSON.stringify(o));
    }
}

interface ILogger { log(msg: string): void; }
interface IDatabase { save(payload: string): void; }

class CleanOrderService {
    constructor(private logger: ILogger, private db: IDatabase) {}
    checkout(o: Order) {
        this.logger.log("Processing " + JSON.stringify(o));
        this.db.save(JSON.stringify(o));
        this.logger.log("Completed " + JSON.stringify(o));
    }
}

class MockLogger implements ILogger { log(msg: string) { console.log("[MockLogger] " + msg); } }
class MockDatabase implements IDatabase { save(payload: string) { console.log("[MockDatabase] saved: " + payload); } }

class EventBus {
    private listeners: Array<(e: any) => void> = [];
    subscribe(fn: (e: any) => void) { this.listeners.push(fn); }
    publish(e: any) { this.listeners.forEach(fn => fn(e)); }
}
class EventOrderService {
    constructor(private bus: EventBus) {}
    checkout(o: Order) { this.bus.publish(o); }
}

function orderHandlerFactory(logger: ILogger, db: IDatabase) {
    return (e: any) => {
        if (!e || !e.id) return;
        logger.log("Event handling " + JSON.stringify(e));
        db.save(JSON.stringify(e));
    };
}

console.log("=== POOR: High CBO (tight coupling) ===");
new OrderService().checkout({ id: "o-1", amount: 99.9 });

console.log("\n=== GOOD: Low CBO (DI via interfaces) ===");
new CleanOrderService(new MockLogger(), new MockDatabase()).checkout({ id: "o-2", amount: 199.9 });

console.log("\n=== GOOD: Adapter for legacy DB ===");
new CleanOrderService(new Logger(), new LegacyDbAdapter()).checkout({ id: "o-3", amount: 299.9 });

console.log("\n=== GOOD: Event-driven decoupling ===");
const bus = new EventBus();
bus.subscribe(orderHandlerFactory(new Logger(), new MockDatabase()));
const evt = new EventOrderService(bus);
evt.checkout({ id: "o-4", amount: 49.9 });

File: 5Days-SoftwareArchitecture/Java/Day-1 CC Metrics/OrderCboDemo.java
import java.util.ArrayList;
import java.util.List;

class Logger {
    void log(String msg) { System.out.println("[Logger] " + msg); }
}

class SqlDatabase {
    void save(String payload) { System.out.println("[SqlDatabase] saved: " + payload); }
}

class LegacyDb {
    void persist(String payload) { System.out.println("[LegacyDb] persisted: " + payload); }
}

class LegacyDbAdapter extends SqlDatabase {
    private final LegacyDb legacy = new LegacyDb();
    @Override
    void save(String payload) { legacy.persist(payload); }
}

class Order {
    final String id;
    final double amount;
    Order(String id, double amount) { this.id = id; this.amount = amount; }
    public String toString() { return id + ":" + amount; }
}

class OrderService {
    void checkout(Order o) {
        Logger logger = new Logger();
        SqlDatabase db = new SqlDatabase();
        logger.log("Processing " + o);
        db.save(o.toString());
        logger.log("Completed " + o);
    }
}

interface ILogger { void log(String msg); }
interface IDatabase { void save(String payload); }

class CleanOrderService {
    private final ILogger logger;
    private final IDatabase db;
    CleanOrderService(ILogger logger, IDatabase db) { this.logger = logger; this.db = db; }
    void checkout(Order o) {
        logger.log("Processing " + o);
        db.save(o.toString());
        logger.log("Completed " + o);
    }
}

class MockLogger implements ILogger { public void log(String msg) { System.out.println("[MockLogger] " + msg); } }
class MockDatabase implements IDatabase { public void save(String payload) { System.out.println("[MockDatabase] saved: " + payload); } }

interface EventListener { void on(Object event); }

class EventBus {
    private final List<EventListener> listeners = new ArrayList<>();
    void subscribe(EventListener l) { listeners.add(l); }
    void publish(Object e) { for (EventListener l : listeners) l.on(e); }
}

class EventOrderService {
    private final EventBus bus;
    EventOrderService(EventBus bus) { this.bus = bus; }
    void checkout(Order o) { bus.publish(o); }
}

class OrderHandler implements EventListener {
    private final ILogger logger;
    private final IDatabase db;
    OrderHandler(ILogger logger, IDatabase db) { this.logger = logger; this.db = db; }
    public void on(Object event) {
        if (!(event instanceof Order)) return;
        Order o = (Order) event;
        logger.log("Event handling " + o);
        db.save(o.toString());
    }
}

public class OrderCboDemo {
    public static void main(String[] args) {
        System.out.println("=== POOR: High CBO (tight coupling) ===");
        OrderService poor = new OrderService();
        poor.checkout(new Order("o-1", 99.9));

        System.out.println("\n=== GOOD: Low CBO (DI via interfaces) ===");
        CleanOrderService clean = new CleanOrderService(new MockLogger(), new MockDatabase());
        clean.checkout(new Order("o-2", 199.9));

        System.out.println("\n=== GOOD: Adapter for legacy DB (still low CBO) ===");
        CleanOrderService adapted = new CleanOrderService(new Logger()::log, new LegacyDbAdapter()::save);
        adapted.checkout(new Order("o-3", 299.9));

        System.out.println("\n=== GOOD: Event-driven decoupling ===");
        EventBus bus = new EventBus();
        OrderHandler handler = new OrderHandler(new Logger()::log, new MockDatabase());
        bus.subscribe(handler);
        EventOrderService evtService = new EventOrderService(bus);
        evtService.checkout(new Order("o-4", 49.9));
    }
}
